lib/common/format.cc

```diff
@@ -0,0 +1,620 @@
+    case 'C': Advance(TokenKind::DC); break;
+    case 'P': Advance(TokenKind::DP); break;
+    case 'T': Advance(TokenKind::DT); break;
+    default: token_.set_kind(TokenKind::D); break;
+    }
+    break;
+  case 'E':
+    switch (LookAheadChar()) {
+    case 'N': Advance(TokenKind::EN); break;
+    case 'S': Advance(TokenKind::ES); break;
+    case 'X': Advance(TokenKind::EX); break;
+    default: token_.set_kind(TokenKind::E); break;
+    }
+    break;
+  case 'F': token_.set_kind(TokenKind::F); break;
+  case 'G': token_.set_kind(TokenKind::G); break;
+  case 'I': token_.set_kind(TokenKind::I); break;
+  case 'L': token_.set_kind(TokenKind::L); break;
+  case 'O': token_.set_kind(TokenKind::O); break;
+  case 'P': token_.set_kind(TokenKind::P); break;
+  case 'R':
+    switch (LookAheadChar()) {
+    case 'C': Advance(TokenKind::RC); break;
+    case 'D': Advance(TokenKind::RD); break;
+    case 'N': Advance(TokenKind::RN); break;
+    case 'P': Advance(TokenKind::RP); break;
+    case 'U': Advance(TokenKind::RU); break;
+    case 'Z': Advance(TokenKind::RZ); break;
+    default: token_.set_kind(TokenKind::None); break;
+    }
+    break;
+  case 'S':
+    switch (LookAheadChar()) {
+    case 'P': Advance(TokenKind::SP); break;
+    case 'S': Advance(TokenKind::SS); break;
+    default: token_.set_kind(TokenKind::S); break;
+    }
+    break;
+  case 'T':
+    switch (LookAheadChar()) {
+    case 'L': Advance(TokenKind::TL); break;
+    case 'R': Advance(TokenKind::TR); break;
+    default: token_.set_kind(TokenKind::T); break;
+    }
+    break;
+  case 'X': token_.set_kind(TokenKind::X); break;
+  case 'Z': token_.set_kind(TokenKind::Z); break;
+  case '-':
+  case '+': token_.set_kind(TokenKind::Sign); break;
+  case '/': token_.set_kind(TokenKind::Slash); break;
+  case '(': token_.set_kind(TokenKind::LParen); break;
+  case ')': token_.set_kind(TokenKind::RParen); break;
+  case '.': token_.set_kind(TokenKind::Point); break;
+  case ':': token_.set_kind(TokenKind::Colon); break;
+  case '\\': token_.set_kind(TokenKind::Backslash); break;
+  case '$': token_.set_kind(TokenKind::Dollar); break;
+  case '*':
+    token_.set_kind(LookAheadChar() == '(' ? TokenKind::Star : TokenKind::None);
+    break;
+  case ',': {
+    token_.set_kind(TokenKind::Comma);
+    CHAR laChar = LookAheadChar();
+    if (laChar == ',') {
+      Advance(TokenKind::Comma);
+      token_.set_offset(cursor_ - format_);
+      AppendError("Unexpected ',' in format expression");
+    } else if (laChar == ')') {
+      AppendError("Unexpected ',' before ')' in format expression");
+    }
+    break;
+  }
+  case '\'':
+  case '"':
+    for (++cursor_; cursor_ < end_; ++cursor_) {
+      if (*cursor_ == c) {
+        if (auto nc{cursor_ + 1}; nc < end_ && *nc != c) {
+          token_.set_kind(TokenKind::String);
+          break;
+        }
+        ++cursor_;
+      }
+    }
+    token_.set_length(cursor_ - format_ - token_.offset() + (cursor_ < end_));
```

Comment:
Adding a bool to an integer. Might want to consider a ternary (?1:0) to make the intent explicit.

Reply:
Ok:

 token_.set_length(
        cursor_ - format_ - token_.offset() + (cursor_ < end_ ? 1 : 0));

Reply:
The language explicitly defines `true` as being 1 when cast to an integral type.  It's probably less ugly without the redundant reimplementation of the conversion, IMO.

Reply:
I'd add a new statement with an 'if' that adds one.  It's clear and not ugly.  The boolean-as-integer expr inline is hard to see.

Reply:
Ok.  My mild preference is for the original.  I don't mind the ternary op too much, and I understand the point of the suggestion.  I'll switch back.

Reply:
Of these suggestions, my preference is for the original.  If I get two votes for something else, or another suggestion that looks better, I'll go with it.